### PR TITLE
Add checking just before creating bucket, ensuring no mp upload is alive

### PR DESCRIPTION
**Pull request by kuenishi**
addresses https://github.com/basho/riak_cs/issues/475  with https://github.com/basho/riak_cs/pull/842 .

### CODE REVIEW

**Comment by kuenishi** [comment]
Additional todo: add special cleanup api to stanchion, to compensate operation around https://github.com/basho/riak_cs/pull/842

File: src/stanchion_multipart.erl [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% ---------------------------------------------------------------------
+
+-module(stanchion_multipart).
+
+-export([check_no_multipart_uploads/2]).
+
+%%　-include("riak_moss.hrl").
```
Comment (shino) [inline src/stanchion_multipart.erl]: Full-width space just after `%%`. :space_invader:

File: src/stanchion_multipart.erl [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+-include("stanchion.hrl").
+
+-type cluster_id() :: undefined | term().  % Type still in flux.
```
Comment (shino) [inline src/stanchion_multipart.erl]: `cluster_id` is `undefined | binary()` in `riak_cs.hrl`, and I believe so :)

File: src/stanchion_multipart.erl [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    %% check all up
+    lists:all(fun(Key) ->
+                      has_any_upload(HashBucket, Key, RiakPid)
+              end, Keys).
+
+has_any_upload(HashBucket, Key, RiakPid) ->    
+    {ok, Obj} = riakc_pb_socket:get(RiakPid, HashBucket, Key),
+    Manifests = manifests_from_riak_object(Obj),
+    lists:all(fun({_UUID,Manifest}) ->
+                      case Manifest#lfs_manifest_v3.state of
```
Comment (shino) [inline src/stanchion_multipart.erl]: Nitpick. Is `?MANIFEST` better instread of `#lfs_manifest_v3`?

File: src/stanchion_multipart.erl [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                          writing -> 
+                              %% if this is mp => false
+                              not proplists:is_defined(multipart, Manifest#lfs_manifest_v3.props);
```
Comment (shino) [inline src/stanchion_multipart.erl]: ditto

File: src/stanchion_multipart.erl [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+check_no_multipart_uploads(Bucket, RiakPid) ->
+    HashBucket = stanchion_utils:to_bucket_name(objects, Bucket),
+    {ok, Keys} = riakc_pb_socket:list_keys(RiakPid, HashBucket),
+    
+    %% check all up
+    lists:all(fun(Key) ->
+                      has_any_upload(HashBucket, Key, RiakPid)
+              end, Keys).
+
+has_any_upload(HashBucket, Key, RiakPid) ->
```
Comment (shino) [inline src/stanchion_multipart.erl]: Does this function return `true` when _no_ MP uploads which are initiated but not completed?
Its name should be something like `has_no_uploads` ?

File: src/stanchion_utils.erl [hunk 6]
```diff
@@ -551,18 +552,17 @@ bucket_available(Bucket, RequesterId, BucketOp, RiakPid) ->
                  BucketOp == update_acl) ->
                     {false, no_such_bucket};
                 (OwnerId == RequesterId andalso
-                BucketOp == create)
+                 BucketOp == create) ->
+                    is_bucket_ready_to_create(Bucket, RiakPid, BucketObj),
```
Comment (shino) [inline src/stanchion_utils.erl]: Where have the return value gone?

File: src/stanchion_utils.erl [hunk 7]
```diff
@@ -620,6 +620,39 @@ do_bucket_op(Bucket, OwnerId, AclOrPolicy, BucketOp) ->
             Else
     end.
 
+%% @doc bucket is ok to delete when bucket is empty. Ongoing multipart
+%% uploads are all supposed to be automatically aborted by Riak CS.
+%% If the bucket still has active objects, just fail. Else if the
+%% bucket still has ongoing multipart, Stanchion returns error and
+%% Riak CS retries sometimes, in case of concurrent multipart
```
Comment (shino) [inline src/stanchion_utils.erl]: s/sometimes/some times/  (Dictionary says "sometimes" is a synonym of "occasionally", right?)

File: src/stanchion_utils.erl [hunk 8]
```diff
@@ -620,6 +620,39 @@ do_bucket_op(Bucket, OwnerId, AclOrPolicy, BucketOp) ->
...
+
+%% @doc ensure there are no multipart uploads in creation time because
+%% multipart uploads remains in deleted buckets in former versions
+%% before 1.5.0 (or 1.4.6) where the bug (identified in riak_cs/#475).
+is_bucket_ready_to_create(Bucket, RiakPid, BucketObj) ->
+    is_bucket_clean(Bucket, RiakPid, BucketObj).
+
+is_bucket_clean(Bucket, RiakPid, BucketObj) ->
+    %% for debugging CS
+    %% {false, remaining_multipart_upload}.
```
Comment (shino) [inline src/stanchion_utils.erl]: Do you mind deleting these two lines?

File: src/stanchion_multipart.erl [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+    %% Upgrade the manifests to be the latest erlang
+    %% record version
+    Upgraded = upgrade_wrapped_manifests(DecodedSiblings),
+
+    %% resolve the siblings
+    Resolved = stanchion_manifest_resolution:resolve(Upgraded),
+
+    %% prune old scheduled_delete manifests
+    %%stanchion_manifest_utils:prune(Resolved).
```
Comment (shino) [inline src/stanchion_multipart.erl]: To be commented out or to be deleted...

**Comment by kuenishi** [comment]
Discussion concluded that cleanup tool will be too big to fry at once, thus concluded:
- add the cleanup tool will be written in Riak CS repo
- which has an option to specify timestamp, which deletes manifests older than that timestamp
- Before operation, users need to calibrate all clocks in the cluster, at the minutes level (no need to exact accuracy). Just mention it in the release notes

**Comment by shino** [comment]
`list_keys` are executed twice (to check no active objects and to check no mp initiated writing ones).

We could change these to single listing or streaming list keys or fold objects.
Rewriting is deferred now and out of the scope of this PR.

**Comment by kuenishi** [comment]
@shino ready to review again :)

File: src/stanchion_utils.erl [hunk 10]
```diff
@@ -544,25 +544,23 @@ bucket_available(Bucket, RequesterId, BucketOp, RiakPid) ->
...
                 (BucketOp == delete
                  orelse
                  BucketOp == update_acl) ->
                     {false, no_such_bucket};
                 (OwnerId == RequesterId andalso
-                BucketOp == create)
+                 BucketOp == create) ->
+                    is_bucket_ready_to_create(Bucket, RiakPid, BucketObj);
+                (OwnerId == RequesterId andalso
+                 BucketOp == create)
```
Comment (shino) [inline src/stanchion_utils.erl]: The condition is the same as at lines 553-554?

File: src/stanchion_utils.erl [hunk 11]
```diff
@@ -544,25 +544,23 @@ bucket_available(Bucket, RequesterId, BucketOp, RiakPid) ->
...
+                    is_bucket_ready_to_create(Bucket, RiakPid, BucketObj);
                 OwnerId == ?FREE_BUCKET_MARKER andalso
                 (BucketOp == delete
                  orelse
                  BucketOp == update_acl) ->
                     {false, no_such_bucket};
                 (OwnerId == RequesterId andalso
-                BucketOp == create)
+                 BucketOp == create) ->
+                    is_bucket_ready_to_create(Bucket, RiakPid, BucketObj);
```
Comment (shino) [inline src/stanchion_utils.erl]: Owner requests creating bucket, so no check needed?

File: src/stanchion_utils.erl [hunk 12]
```diff
@@ -544,25 +544,23 @@ bucket_available(Bucket, RequesterId, BucketOp, RiakPid) ->
...
+                (OwnerId == RequesterId andalso
+                 BucketOp == create)
                 orelse
                 BucketOp == update_acl ->
```
Comment (shino) [inline src/stanchion_utils.erl]: Should the "orelse" be combined with "BucketOp==create"?

File: src/stanchion_utils.erl [hunk 13]
```diff
@@ -544,25 +544,23 @@ bucket_available(Bucket, RequesterId, BucketOp, RiakPid) ->
             if
                 OwnerId == ?FREE_BUCKET_MARKER andalso
                 BucketOp == create ->
-                    {true, BucketObj};
+                    is_bucket_ready_to_create(Bucket, RiakPid, BucketObj);
                 OwnerId == ?FREE_BUCKET_MARKER andalso
                 (BucketOp == delete
                  orelse
                  BucketOp == update_acl) ->
```
Comment (shino) [inline src/stanchion_utils.erl]: (Not related to this PR...)  `updale_policy` also goes here?

File: src/stanchion_utils.erl [hunk 14]
```diff
@@ -620,6 +603,46 @@ do_bucket_op(Bucket, OwnerId, AclOrPolicy, BucketOp) ->
...
+%% @doc bucket is ok to delete when bucket is empty. Ongoing multipart
+%% uploads are all supposed to be automatically aborted by Riak CS.
+%% If the bucket still has active objects, just fail. Else if the
+%% bucket still has ongoing multipart, Stanchion returns error and
+%% Riak CS retries some times, in case of concurrent multipart
+%% initiation occuring.  After a few retry Riak CS will eventually
+%% returns error to the client (maybe 500?)  Or fallback to heavy
+%% abort-all-multipart and then deletes bucket?  This will be a big
+%% TODO.
+-spec is_bucket_ready_to_delete(binary(), pid(), riakc_obj()) -> {boolean(), riakc_obj()}.
```
Comment (shino) [inline src/stanchion_utils.erl]: If the first element is `false` then second is `atom()` (as the spec of `bucket_availaty` says).

File: src/stanchion_utils.erl [hunk 15]
```diff
@@ -620,6 +603,46 @@ do_bucket_op(Bucket, OwnerId, AclOrPolicy, BucketOp) ->
...
+is_bucket_ready_to_delete(Bucket, RiakPid, BucketObj) ->
+    is_bucket_clean(Bucket, RiakPid, BucketObj).
+
+%% @doc ensure there are no multipart uploads in creation time because
+%% multipart uploads remains in deleted buckets in former versions
+%% before 1.5.0 (or 1.4.6) where the bug (identified in riak_cs/#475).
+-spec is_bucket_ready_to_create(binary(), pid(), riakc_obj()) -> {boolean(), riakc_obj()}.
```
Comment (shino) [inline src/stanchion_utils.erl]: Ditto.

File: src/stanchion_utils.erl [hunk 16]
```diff
@@ -620,6 +603,46 @@ do_bucket_op(Bucket, OwnerId, AclOrPolicy, BucketOp) ->
...
+is_bucket_ready_to_create(Bucket, RiakPid, BucketObj) ->
+    is_bucket_clean(Bucket, RiakPid, BucketObj).
+
+%% @doc here runs two list_keys, one in bucket_empty/2, another in
+%% stanchion_multipart:check_no_multipart_uploads/2. If there are
+%% bunch of pending_delete manifests this may slow (twice as before
+%% #475 fix). If there are bunch of scheduled_delete manifests, this
+%% may also slow, but wait for Garbage Collection to collect those
+%% trashes may improve the speed. => TODO.
+-spec is_bucket_clean(binary(), pid(), riakc_obj()) -> {boolean(), riakc_obj()}.
```
Comment (shino) [inline src/stanchion_utils.erl]: Ditto.

File: src/stanchion_multipart.erl [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+    HashBucket = stanchion_utils:to_bucket_name(objects, Bucket),
+    {ok, Keys} = riakc_pb_socket:list_keys(RiakPid, HashBucket),
+
+    %% check all up
+    lists:all(fun(Key) ->
+                      has_no_upload(HashBucket, Key, RiakPid)
+              end, Keys).
+
+has_no_upload(HashBucket, Key, RiakPid) ->
+    {ok, Obj} = riakc_pb_socket:get(RiakPid, HashBucket, Key),
```
Comment (shino) [inline src/stanchion_multipart.erl:195]: Do we care about `{error, notfound}`? (cf: `bucket_empty` handles `notfound`)
The `Key` here can be tombstoned one, so it is possible to have `notfoud` by GET.

Comment (kuenishi) [inline src/stanchion_multipart.erl:195]: Fixed in #78 . see https://github.com/basho/stanchion/pull/78/files#diff-7295a1ae0c41a0cd6801658ab10dde21R205

**Comment by kuenishi** [comment]
Squashed to #78 .

**Pull request closed without merging**